### ISSUE TITLE
ngPost ignores server-returned Message ID

### ISSUE THREAD

**Issue by AssRap3r** [mbruel/ngPost#44]
Issue: ngPost doesn't acknowledge (or ignores) when a server returns a different message ID than the one sent. Attempting to use the nzb created will fail because the message id's are wrong. 

Error from Nyuu:
[WARN] Server returned a different Message-ID (5ed02ecc$0$7332$f56682d5@speedium.nl) to the one we sent it (GaBsByCtMrPwErMfAwLcCoCw-1590701769791@nyuu); replacing our ID with received ID.

Nyuu uploaded file (downloads okay)
```
<segment bytes="739660" number="1">5ed02cbc$0$22532$f56682d5@speedium.nl</segment>
<segment bytes="739747" number="2">5ed02cbc$0$7362$f56682d5@speedium.nl</segment>
```

ngPost uploaded file (fails)
```
<segment bytes="716800" number="1">a1ac1dff10b54cd18642475a60850522@ngPost</segment>
<segment bytes="716800" number="2">b059409c6e3345bc8802b4ef39d813b2@ngPost</segment>
```

Steps to reproduce:
Upload a file with ngPost to Speedium.nl provider
Download same file with a different host, ie newsdemon

**Comment by mbruel** [comment]
Hi
are you on Linux? 
Could you try something for me? it's easy, just a telenet log of posting 2 simple Articles.
I'm going to write to them to ask them an access so I can test what happen.

**Comment by AssRap3r** [comment]
I'm using linux, if you can provide the commands needed I can try.  speedium is free for now (beta test ending this week) so you can sign up for an account, posting is enabled by default: https://speedium.nl/sign-up/?lang=en

this is the reason according to their team:

```
Nvyous05/24/2020
basically nyuu supplies shit message-id. We can accept it,but then it migh
t not propagate across usenet.
If we rewrite the ,message ID we solve that problem.
Basically its service on our side 8-)
```
```
Speedium - Hidde05/24/2020
>>>why not let keep people custom msg ids ?
@<msg-id@x.y.z> 
Because not all peers accept those custom/wrong message ID's and the posting won't be available on all backbones
```

**Comment by Tensai75** [comment]
I am currently discussion this issue with the SpeediumDev. Peering should not be a problem with the message-ids generated by ngPost (or also nyuu). I think they have generally activated rewriting of message-ids and I hope I can convince them to deactivate it ;-)

**Comment by mbruel** [comment]
I've created an account.
It's ok, I could easily support rewriting of message-ids.
I'll check it this weekend.

**Comment by mbruel** [comment]
```
post
340 345 <5ed10f21$0$7342$f56682d5@speedium.nl> (forced)
From: test@ngPost.com
Newsgroups: alt.binaries.test
Subject: test
Message-ID: <3986d6554bed7b2b06b5dfd45ed10e8f@ngPost>

simple test
.
240 <5ed10f21$0$7342$f56682d5@speedium.nl> Article posted

post
340 345 <5ed10f42$0$7342$f56682d5@speedium.nl> (forced)
From: test@ngPost.com
Newsgroups: alt.binaries.test
Subject: test
Message-ID: <3986d6554bed7b2b06b5dfd45ed10e8f@ngPost>

simple test
.
240 <5ed10f42$0$7342$f56682d5@speedium.nl> Article posted
quit
205 Transferred 240 bytes in 0 articles, 0 groups.  Disconnecting.
Connection closed by foreign host.
```

It's fine, I'll cope with it in the next release this weekend ;)
basically they accept everything, even if we post twice a same article...

**Comment by Tensai75** [comment]
> basically they accept everything, even if we post twice a same article...

Providing unique message-ids is in the responsibility of the upload program. Message-ids are not checked for duplicate ids by the servers.

Speedium currently rewrites all message-ids by default. But they will deactivate the forced message-id rewriting in the next maintenance window.

**Comment by mbruel** [comment]
> Providing unique message-ids is in the responsibility of the upload program. Message-ids are not checked for duplicate ids by the servers.

Who says that?
It makes sense for a provider to create the Message-IDs. This makes SURE that there won't be any duplicates.
it's also faster as they don't need to check if a Message-ID already exist.
I think they should keep it.

**Comment by mbruel** [comment]
> Providing unique message-ids is in the responsibility of the upload program. Message-ids are not checked for duplicate ids by the servers.

You're wrong, of course there is a check on the provider side for the uniqueness of the Message-IDs cause they reject existing ones if they don't rewrite them!


```
post
340 Send Article to be Posted
From: test@ngPost.com
Newsgroups: alt.binaries.test
Subject: test
Message-ID: <3986d6554bed7b2b06b5dfd45ed10e8f@ngPost>

simple test
.

240 Article Posted

post
340 Send Article to be Posted
From: test@ngPost.com
Newsgroups: alt.binaries.test
Subject: test
Message-ID: <3986d6554bed7b2b06b5dfd45ed10e8f@ngPost>

simple test
.
441 Posting Failed. Message-ID is not unique E1

```

**Comment by mbruel** [comment]
Alright, I've done the change in the code.
If you guys compile from the sources you can test it.
I'll probably work on few other things before doing the new release.

Here is an nzb posted on speedium
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE nzb PUBLIC "-//newzBin//DTD NZB 1.1//EN" "http://www.newzbin.com/DTD/nzb/nzb-1.1.dtd">
<nzb xmlns="http://www.newzbin.com/DTD/2003/nzb">
  <file poster="adbOHlVS7TTy4@ngPost.com" date="1590782625" subject="[1/1] - &quot;Odin_3.09.3.zip&quot; yEnc (1/2) 1004682">
    <groups>
      <group>alt.binaries.test</group>
    </groups>
    <segments>
      <segment bytes="716800" number="1">5ed16aa0$0$7356$f56682d5@speedium.nl</segment>
      <segment bytes="287882" number="2">5ed16aa0$1$7356$f56682d5@speedium.nl</segment>
    </segments>
  </file>
</nzb>
```

@AssRap3r please close the issue if you confirm it's working now ;)
Let me know if you need an AppImage

**Comment by mbruel** [comment]
How did you guys contacted the dev team?

**Comment by Tensai75** [comment]
> How did you guys contacted the dev team?

You can contact them via their Discord Channel: https://discord.gg/2hxXZBS

**Comment by mbruel** [comment]
I've never used Discord...
When I connect it tells me "You do not have permission to send messages in this channel"...

**Comment by AssRap3r** [comment]
@mbruel compiled from git source but not seeing any changes, nzb is still output with @ngPost id

**Comment by mbruel** [comment]
@AssRap3r arf.. my fault, I forgot to commit the actual implementation in NntpConnection...
Try now, it should be fine. :)

Please confirm and if all good close the issue ;)

**Comment by mbruel** [comment]
if you use "--fulldebug" instead of simple "-d" you should see this kind of log:

`[Poster #1 {NntpCon #2}] the server has overwritten the Message-ID to : 5ed1bec6$0$403$f56682d5@speedium.nl`

Within the GUI, you need to set up the debug level to 2 to see that message.

**Comment by AssRap3r** [comment]
@mbruel compiled newest, there are still random instances where ngPost is present

1st try: Found 33 occurrence(s) in 1 file(s) (800MB test)
2nd try: Found 8 occurrence(s) in 1 file(s) (800MB test)
3rd try: OK (800MB test)
4th try: OK (800MB test)
5th try: Found 8 occurrence(s) in 1 file(s) (5GB test)


```
<segment bytes="716800" number="117">5ed22c3e$2$512$f56682d5@speedium.nl</segment>
<segment bytes="716800" number="118">b30b79e29f4c4a72915b5bb652d16d12@ngPost</segment>
<segment bytes="716800" number="119">5ed22c3e$2$34171$f56682d5@speedium.nl</segment>
```
```
[Poster #1 {NntpCon #6}] start sending article: [45/51] 10GB.part045.rar - Article #117/147 <id: 1109d615f15f4e61bc13381beba9d56b, nbTrySend: 0>
[Poster #1 {NntpCon #6}] the server has overwritten the Message-ID to : 5ed22c3e$2$512$f56682d5@speedium.nl (article: 1109d615f15f4e61bc13381beba9d56b@ngPost)

[Poster #0 {NntpCon #4}] start sending article: [45/51] 10GB.part045.rar - Article #118/147 <id: b30b79e29f4c4a72915b5bb652d16d12, nbTrySend: 0>

[Poster #0 {NntpCon #2}] start sending article: [45/51] 10GB.part045.rar - Article #119/147 <id: 25328784bef040c8bd91b09d07e83548, nbTrySend: 0>
[Poster #0 {NntpCon #2}] the server has overwritten the Message-ID to : 5ed22c3e$2$34171$f56682d5@speedium.nl (article: 25328784bef040c8bd91b09d07e83548@ngPost)


```

**Comment by mbruel** [comment]
@AssRap3r 
Hi man,
That's weird, would it mean they didn't overwrite those...?
I'm using C style check to find first the `<` then the `>`. It should be robust...

I've added extra debug information in the case you use Speedium and there is NO overwrite of the Message-ID. I'm printing out their full answer.
You should see stuff like that:
`[DEBUG_Speedium]< No overwrite for article %1, server resp: %2`
or 
`[DEBUG_Speedium]> No overwrite for article %1, server resp: %2`

You don't need to launch in debug, you can use the release and only -d so you won't get too many traces.

Could you pull, compile in release and try again so with -d or --debug. Please redirect the whole output, either with a `> log.txt` or using `| tee log.txt` at the end of your command.

Thanks for your help and test ;)

**Comment by AssRap3r** [comment]
[speedium-test-log.txt](https://github.com/mbruel/ngPost/files/4705578/speedium-test-log.txt)
[speedium-test.zip](https://github.com/mbruel/ngPost/files/4705579/speedium-test.zip) (nzb)

[DEBUG_Speedium]< No overwrite for article e008ad701b4d40e68b25f3f93d79c50d@ngPost, server resp: 240  Article posted

Found 94 occurrence(s) in 1 file(s)
Found 51 files and 7497 segments totaling 4.99 GB
Available: 7407 [98.80%], Missing:   90 [1.20%], Total: 7497 [100.00%]

**Comment by mbruel** [comment]
Ok then it seems, they don't overwrite ALL the Articles which is quite weird.
The dev team of Speedium is in weekend, so I'm going to do.
I'll chat with them on Monday.
Thanks.
What is your "Available: 7407 [98.80%], Missing: 90 [1.20%], Total: 7497 [100.00%]"
Are you trying to download on the same provider or another one?

PS: to avoid the progress bar disturbance, use `-d --disp_progress FILES` next time.

**Comment by AssRap3r** [comment]
@mbruel trying to download on a different provider (Newsdemon). the 98% means it's partially working but as you said the remaining articles still tagged as ngPost are speedium's problem. hopefully they can sort it.

**Comment by mbruel** [comment]
and if you download from speedium, the problem is the same I guess?
I'm on a chat with the dev team, we'll discuss that on Monday.
Thanks ;)

**Comment by AssRap3r** [comment]
yes, finds speedium appended but misses ngPost

**Comment by mbruel** [comment]
Yep I guess we can close the issue.
I'm going to let the debug traces for some time to help speedium guys to debug their issue next week.
I'll do the v4.8 by the end of next week.
Thanks for raising the issue, that's definitely something that was missing in ngPost ;)
